Issue with listing data sets after opening with encoding

<!--
  Before opening a new issue, please search our existing issues: https://github.com/zowe/zowe-explorer-vscode/issues
-->

**Describe the bug**

If I open a data set in a specific encoding, then provide a search/filter for a data set that does not exist, then enter a filter that should result in data sets being listed, I get `TypeError: Cannot read properties of undefined (reading 'startsWith')`

**To Reproduce**

1. Open a data set with encoding
2. Enter a new filter that will not return any results (i.e., enter a filter that would not find any data sets/members)
3. Enter a new filter that **should** return results
4. Get error message

**Expected behavior**

I should see a list of data sets.

**Screenshots**

![EncodingIssue](https://github.com/user-attachments/assets/65e4f0e3-dcb4-4e2b-b4d5-25c14683fc66)

**Desktop (please complete the following information):**

- OS: Win 10
- Zowe Explorer Version: 3.0.3 and latest build of 3.1.0 pre-release
- (Optional) Zowe CLI Version:
- (Optional) Are you using Secure Credential Store?
<!-- If you have SCS installed, please read the documentation for Integrating SCS within Zowe Explorer: https://github.com/zowe/zowe-explorer-vscode/blob/main/docs/README-Security.md -->

**Additional context**

More complete logs (produced using 3.0.3):

Opening filter menu:
```
[2024/12/05 09:40:24] [TRACE] DatasetTree.filterPrompt called.
[2024/12/05 09:40:24] [TRACE] DatasetTree.datasetFilterPrompt called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.checkCurrentProfile called.
[2024/12/05 09:40:24] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:24] [TRACE] Profiles.checkCurrentProfile called.
[2024/12/05 09:40:24] [TRACE] Profiles.getProfileInfo called.
[2024/12/05 09:40:24] [TRACE] Profiles.getProfileSetting called.
[2024/12/05 09:40:24] [TRACE] Profiles.validateProfiles called.
[2024/12/05 09:40:24] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:24] [TRACE] Profiles.getProfileInfo called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.nodeDataChanged called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.nodeDataChanged called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.nodeDataChanged called.
[2024/12/05 09:40:24] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.refresh called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:24] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:24] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:24] [DEBUG] Prompting the user for a data set pattern
[2024/12/05 09:40:24] [TRACE] PersistentFilters.getSearchHistory called.
[2024/12/05 09:40:24] [TRACE] PersistentFilters.getSearchHistory called.
[2024/12/05 09:40:24] [TRACE] DatasetTree.getChildren called.
[2024/12/05 09:40:24] [TRACE] ZoweDatasetNode.getChildren called for lpar1.zosmf.
[2024/12/05 09:40:24] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:24] [TRACE] ZoweDatasetNode.getDatasets called.
[2024/12/05 09:40:25] [TRACE] ZoweDatasetNode.setEncoding called.
[2024/12/05 09:40:25] [TRACE] DatasetTree.getChildren called.
[2024/12/05 09:40:25] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:25] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:25] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:25] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:25] [TRACE] ZoweTreeProvider.getTreeItem called.
[2024/12/05 09:40:25] [TRACE] DatasetTree.getChildren called.
[2024/12/05 09:40:25] [TRACE] ZoweDatasetNode.getChildren called for lpar1.zosmf.
[2024/12/05 09:40:25] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:25] [TRACE] ZoweDatasetNode.getDatasets called.
[2024/12/05 09:40:25] [TRACE] ZoweDatasetNode.setEncoding called.
```
Selecting a filter and executing the search:
```
[2024/12/05 09:40:56] [TRACE] ProfilesUtils.syncSessionNode called.
[2024/12/05 09:40:56] [TRACE] DatasetTree.getChildren called.
[2024/12/05 09:40:56] [TRACE] ZoweDatasetNode.getChildren called for lpar1.zosmf.
[2024/12/05 09:40:56] [TRACE] Profiles.getInstance called.
[2024/12/05 09:40:56] [TRACE] ZoweDatasetNode.getDatasets called.
[2024/12/05 09:40:56] [TRACE] ZoweDatasetNode.setEncoding called.
[2024/12/05 09:40:56] [ERROR] TypeError: Cannot read properties of undefined (reading 'startsWith')
{
  errorDetails: [TypeError: Cannot read properties of undefined (reading 'startsWith')
  	at Fe.setEncoding (c:\Users\me\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0-SNAPSHOT\out\src\main.extension.js:64:24000)
  	at new Fe (c:\Users\me\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0-SNAPSHOT\out\src\main.extension.js:64:12388)
  	at Fe.<anonymous> (c:\Users\me\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0-SNAPSHOT\out\src\main.extension.js:64:17361)
  	at Generator.next (<anonymous>)
  	at r (c:\Users\me\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0-SNAPSHOT\out\src\main.extension.js:64:11949)
  	at process.processTicksAndRejections (node:internal/process/task_queues:95:5)],
  apiType: 'mvs',
  profile: undefined
}
```

--- a/packages/zowe-explorer/CHANGELOG.md
+++ b/packages/zowe-explorer/CHANGELOG.md
@@ -28,6 +28,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 - Fixed an issue where a migrated data set is unusable after it is recalled through Zowe Explorer. [#3294](https://github.com/zowe/zowe-explorer-vscode/issues/3294)
 - Fixed an issue where a recalled PDS is expandable after it is migrated through Zowe Explorer. [#3294](https://github.com/zowe/zowe-explorer-vscode/issues/3294)
 - Fixed an issue where data set nodes did not update if migrated or recalled outside of Zowe Explorer. [#3294](https://github.com/zowe/zowe-explorer-vscode/issues/3294)
+- Fixed an issue where listing data sets resulted in an error after opening a data set with an encoding. [#3347](https://github.com/zowe/zowe-explorer-vscode/issues/3347)
 
 ## `3.0.3`
 

--- a/packages/zowe-explorer/__tests__/__unit__/trees/dataset/DatasetTree.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/trees/dataset/DatasetTree.unit.test.ts
@@ -3297,30 +3297,31 @@ describe("Dataset Tree Unit Tests - Sorting and Filtering operations", () => {
 
 describe("Dataset Tree Unit Tests - Function openWithEncoding", () => {
     it("sets binary encoding if selection was made", async () => {
-        const setEncodingMock = jest.spyOn(DatasetFSProvider.instance, "setEncodingForFile").mockImplementation();
         const node = new ZoweDatasetNode({ label: "encodingTest", collapsibleState: vscode.TreeItemCollapsibleState.None });
+        const setEncodingMock = jest.spyOn(node, "setEncoding").mockImplementation();
         node.openDs = jest.fn();
         jest.spyOn(SharedUtils, "promptForEncoding").mockResolvedValueOnce({ kind: "binary" });
         await DatasetTree.prototype.openWithEncoding(node);
-        expect(setEncodingMock).toHaveBeenCalledWith(node.resourceUri, { kind: "binary" });
+        expect(setEncodingMock).toHaveBeenCalledWith({ kind: "binary" });
         expect(node.openDs).toHaveBeenCalledTimes(1);
         setEncodingMock.mockRestore();
     });
 
     it("sets text encoding if selection was made", async () => {
-        const setEncodingMock = jest.spyOn(DatasetFSProvider.instance, "setEncodingForFile").mockImplementation();
+        jest.spyOn(DatasetFSProvider.instance, "exists").mockReturnValueOnce(true);
         const node = new ZoweDatasetNode({ label: "encodingTest", collapsibleState: vscode.TreeItemCollapsibleState.None });
+        const setEncodingMock = jest.spyOn(node, "setEncoding").mockImplementation();
         node.openDs = jest.fn();
         jest.spyOn(SharedUtils, "promptForEncoding").mockResolvedValueOnce({ kind: "text" });
         await DatasetTree.prototype.openWithEncoding(node);
-        expect(setEncodingMock).toHaveBeenCalledWith(node.resourceUri, { kind: "text" });
+        expect(setEncodingMock).toHaveBeenCalledWith({ kind: "text" });
         expect(node.openDs).toHaveBeenCalledTimes(1);
         setEncodingMock.mockRestore();
     });
 
     it("does not set encoding if prompt was cancelled", async () => {
-        const setEncodingSpy = jest.spyOn(DatasetFSProvider.instance, "setEncodingForFile");
         const node = new ZoweDatasetNode({ label: "encodingTest", collapsibleState: vscode.TreeItemCollapsibleState.None });
+        const setEncodingSpy = jest.spyOn(node, "setEncoding");
         node.openDs = jest.fn();
         jest.spyOn(SharedUtils, "promptForEncoding").mockResolvedValueOnce(undefined);
         await DatasetTree.prototype.openWithEncoding(node);

--- a/packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts
@@ -658,9 +658,16 @@ describe("ZoweDatasetNode Unit Tests - Function node.openDs()", () => {
 });
 
 describe("ZoweDatasetNode Unit Tests - Function node.setEncoding()", () => {
-    const setEncodingForFileMock = jest.spyOn(DatasetFSProvider.instance, "setEncodingForFile").mockImplementation();
+    let setEncodingForFileMock: jest.SpyInstance;
+    let existsMock: jest.SpyInstance;
+
+    beforeAll(() => {
+        setEncodingForFileMock = jest.spyOn(DatasetFSProvider.instance, "setEncodingForFile").mockImplementation();
+        existsMock = jest.spyOn(DatasetFSProvider.instance, "exists").mockReturnValue(true);
+    });
 
     afterAll(() => {
+        existsMock.mockRestore();
         setEncodingForFileMock.mockRestore();
     });
 

--- a/packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts
@@ -124,9 +124,6 @@ describe("ZoweDatasetNode Unit Tests", () => {
         expect(testNode.getSession()).toBeDefined();
     });
 
-    /*************************************************************************************************************
-     * Creates an ZoweDatasetNode and checks that its members are all initialized by the constructor
-     *************************************************************************************************************/
     it("calls setEncoding when constructing a node with encoding", () => {
         jest.spyOn(BaseProvider.prototype, "setEncodingForFile").mockImplementationOnce(() => {});
         const makeEmptyDsWithEncodingMock = jest.spyOn(DatasetFSProvider.instance, "makeEmptyDsWithEncoding").mockImplementationOnce(() => {});

--- a/packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/trees/dataset/ZoweDatasetNode.unit.test.ts
@@ -10,7 +10,7 @@
  */
 
 import * as vscode from "vscode";
-import { DsEntry, Gui, imperative, PdsEntry, Validation, ZoweScheme } from "@zowe/zowe-explorer-api";
+import { BaseProvider, DsEntry, Gui, imperative, PdsEntry, Validation, ZoweScheme } from "@zowe/zowe-explorer-api";
 import * as zosfiles from "@zowe/zos-files-for-zowe-sdk";
 import {
     createSessCfgFromArgs,
@@ -122,6 +122,29 @@ describe("ZoweDatasetNode Unit Tests", () => {
         expect(testNode.label).toBeDefined();
         expect(testNode.getParent()).toBeUndefined();
         expect(testNode.getSession()).toBeDefined();
+    });
+
+    /*************************************************************************************************************
+     * Creates an ZoweDatasetNode and checks that its members are all initialized by the constructor
+     *************************************************************************************************************/
+    it("calls setEncoding when constructing a node with encoding", () => {
+        jest.spyOn(BaseProvider.prototype, "setEncodingForFile").mockImplementationOnce(() => {});
+        const makeEmptyDsWithEncodingMock = jest.spyOn(DatasetFSProvider.instance, "makeEmptyDsWithEncoding").mockImplementationOnce(() => {});
+        const setEncodingSpy = jest.spyOn(ZoweDatasetNode.prototype, "setEncoding");
+        const testNode = new ZoweDatasetNode({
+            label: "BRTVS99",
+            collapsibleState: vscode.TreeItemCollapsibleState.None,
+            contextOverride: Constants.DS_DS_CONTEXT,
+            encoding: { kind: "binary" },
+            profile: createIProfile(),
+            parentNode: createDatasetSessionNode(createISession(), createIProfile()),
+        });
+
+        expect(testNode.label).toBe("BRTVS99");
+        expect(testNode.collapsibleState).toBe(vscode.TreeItemCollapsibleState.None);
+        expect(setEncodingSpy).toHaveBeenCalled();
+        expect(makeEmptyDsWithEncodingMock).toHaveBeenCalled();
+        setEncodingSpy.mockRestore();
     });
 
     /*************************************************************************************************************

--- a/packages/zowe-explorer/__tests__/__unit__/trees/shared/SharedUtils.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/trees/shared/SharedUtils.unit.test.ts
@@ -528,7 +528,9 @@ describe("Shared utils unit tests - function promptForEncoding", () => {
             parentNode,
             contextOverride: Constants.DS_MEMBER_CONTEXT,
         });
+        const existsMock = jest.spyOn(DatasetFSProvider.instance, "exists").mockReturnValueOnce(true);
         node.setEncoding(otherEncoding);
+        expect(existsMock).toHaveBeenCalled();
         blockMocks.getEncodingForFile.mockReturnValueOnce(undefined);
         await SharedUtils.promptForEncoding(node);
         expect(blockMocks.showQuickPick).toHaveBeenCalled();

--- a/packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts
+++ b/packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts
@@ -81,9 +81,6 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
         } else {
             this.contextValue = isBinary ? Constants.DS_DS_BINARY_CONTEXT : Constants.DS_DS_CONTEXT;
         }
-        if (opts.encoding != null) {
-            this.setEncoding(opts.encoding);
-        }
         this.tooltip = this.label as string;
         const icon = IconGenerator.getIconByNode(this);
         if (icon) {
@@ -134,7 +131,7 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
             }
 
             if (opts.encoding != null) {
-                DatasetFSProvider.instance.makeEmptyDsWithEncoding(this.resourceUri, opts.encoding);
+                this.setEncoding(opts.encoding);
             }
         }
     }
@@ -313,8 +310,8 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
                     } else if (!SharedContext.isMigrated(dsNode) && item.migr?.toUpperCase() === "YES") {
                         dsNode.datasetMigrated();
                     }
-                    // Creates a ZoweDatasetNode for a PDS
                 } else if (item.migr && item.migr.toUpperCase() === "YES") {
+                    // Creates a ZoweDatasetNode for a migrated dataset
                     dsNode = new ZoweDatasetNode({
                         label: item.dsname,
                         collapsibleState: vscode.TreeItemCollapsibleState.None,
@@ -323,8 +320,8 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
                         profile: cachedProfile,
                     });
                     elementChildren[dsNode.label.toString()] = dsNode;
-                    // Creates a ZoweDatasetNode for a VSAM file
                 } else if (item.dsorg === "PO" || item.dsorg === "PO-E") {
+                    // Creates a ZoweDatasetNode for a PDS
                     dsNode = new ZoweDatasetNode({
                         label: item.dsname,
                         collapsibleState: vscode.TreeItemCollapsibleState.Collapsed,
@@ -332,8 +329,8 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
                         profile: cachedProfile,
                     });
                     elementChildren[dsNode.label.toString()] = dsNode;
-                    // Creates a ZoweDatasetNode for a dataset with imperative errors
                 } else if (item.error instanceof imperative.ImperativeError) {
+                    // Creates a ZoweDatasetNode for a dataset with imperative errors
                     dsNode = new ZoweDatasetNode({
                         label: item.dsname,
                         collapsibleState: vscode.TreeItemCollapsibleState.None,
@@ -344,8 +341,8 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
                     dsNode.command = { command: "zowe.placeholderCommand", title: "" };
                     dsNode.errorDetails = item.error; // Save imperative error to avoid extra z/OS requests
                     elementChildren[dsNode.label.toString()] = dsNode;
-                    // Creates a ZoweDatasetNode for a migrated dataset
                 } else if (item.dsorg === "VS") {
+                    // Creates a ZoweDatasetNode for a VSAM file
                     let altLabel = item.dsname;
                     let endPoint = altLabel.indexOf(".DATA");
                     if (endPoint === -1) {
@@ -713,7 +710,11 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
         } else {
             this.contextValue = isMemberNode ? Constants.DS_MEMBER_CONTEXT : Constants.DS_DS_CONTEXT;
         }
-        DatasetFSProvider.instance.setEncodingForFile(this.resourceUri, encoding);
+        if (DatasetFSProvider.instance.exists(this.resourceUri)) {
+            DatasetFSProvider.instance.setEncodingForFile(this.resourceUri, encoding);
+        } else {
+            DatasetFSProvider.instance.makeEmptyDsWithEncoding(this.resourceUri, encoding);
+        }
         const fullPath = isMemberNode ? `${this.getParent().label as string}(${this.label as string})` : (this.label as string);
         if (encoding != null) {
             this.updateEncodingInMap(fullPath, encoding);

--- a/packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts
+++ b/packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts
@@ -71,9 +71,6 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
 
     public constructor(opts: Definitions.IZoweDatasetTreeOpts) {
         super(opts.label, opts.collapsibleState, opts.parentNode, opts.session, opts.profile);
-        if (opts.encoding != null) {
-            this.setEncoding(opts.encoding);
-        }
         const isBinary = opts.encoding?.kind === "binary";
         if (opts.contextOverride) {
             this.contextValue = opts.contextOverride;
@@ -83,6 +80,9 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
             this.contextValue = isBinary ? Constants.DS_MEMBER_BINARY_CONTEXT : Constants.DS_MEMBER_CONTEXT;
         } else {
             this.contextValue = isBinary ? Constants.DS_DS_BINARY_CONTEXT : Constants.DS_DS_CONTEXT;
+        }
+        if (opts.encoding != null) {
+            this.setEncoding(opts.encoding);
         }
         this.tooltip = this.label as string;
         const icon = IconGenerator.getIconByNode(this);

